### PR TITLE
Don't throw exceptions inside OpenMP parallel blocks

### PR DESCRIPTION
**Pull request by colesbury**
Fixes undefined behavior: exceptions are not allowed to be thrown across
OpenMP constructs.

### CODE REVIEW

**Comment by colesbury** [comment]
cc @yf225

**Comment by yf225** [comment]
CPU perf test fail seems to be intermittent. Retesting works: https://ci.pytorch.org/jenkins/job/pytorch-builds/job/short-perf-test-cpu/305/console.

File: aten/src/TH/generic/THTensorMath.c [hunk 1]
```diff
@@ -389,25 +392,34 @@ void THTensor_(take)(THTensor *r_, THTensor *src, THLongTensor *index)
...
+    int64_t idx = index_data[i];
+    if (idx < srcElements && idx >= -srcElements) {
+      idx = THTensor_(wrapLinearIndex)(idx, srcElements);
+      if (isContiguous) {
+        dst_data[i] = src_data[idx];
+      } else {
+        dst_data[i] = src_data[THTensor_(dataOffset)(src, idx)];
+      }
+    } else {
+      THAtomicCompareAndSwapLong(&invalidIdx, -1, idx);
```
Comment (gchanan) [inline aten/src/TH/generic/THTensorMath.c]: I'm curious why you used CAS -- you can just set unconditionally right?  (I don't know if we have that API).

**Review by gchanan — APPROVED** [review]

File: aten/src/TH/generic/THTensorMath.c (hunk 1 above)
Comment (gchanan) [reply aten/src/TH/generic/THTensorMath.c]: I meant set unconditionally but still atomically.

Comment (colesbury) [reply aten/src/TH/generic/THTensorMath.c]: I figured it's slightly better to prefer the first invalid index encountered instead of the last.

**Pull request merged**